### DISCUSSION TITLE
Multi Regional Deployments

### DISCUSSION THREAD

**Discussion by TylerFiekens** [brefphp/bref#1568] [Support]
Hey everyone 👋,

I'm interested in using this framework to deploy my next project.
I am also planning on deploying to multiple regions, for example; eu-west-1 and us-east-1. How would I go about it?
- Does bref support deployment to multiple Regions? 
- How about assets?
- What about global DynamoDB tables for caching?
- What about SQS in multiple regions?

Thanks for the help. 😄

**Comment by mnapoli** [comment]
Hey @TylerFiekens, this is a wide topic (and interesting one), we could work together on it if you are interested.

But just to get the ball rolling for you, some notes: you can deploy the same application to multiple stages with the flag `serverless deploy --region=eu-west-1`. If you use Lift (Bref's official solution for assets), they would be deployed too. However this wouldn't be a "zero-downtime-all-regions-at-once" deployment, but that's impossible anyway.

The resources you want to share across regions (like DynamoDB) you'll want to deploy them separately, and set up global tables. But if it's just for caching, then you might actually not want to do that.

SQS: if these are the same jobs across regions you'll want one queue. If they are separate, you can duplicate across regions. It really depends on what you want to do.

**Reply by TylerFiekens** [reply]
Hey, thanks for the response. I would like to collaborate on this. However i am really not that knowable about AWS (which is probably not that good of an idea when I want to deploy there.), however I know a bit about Laravel, which is the framework I mainly use.

Maybe we could start by defining which components must be globally deployed in order for Laravel to function properly.

### Caching

While I see your point in that a cache does not need to deployed globally I can still see a few use cases which may require this. A) Cache busting - What if user A in region us-east-1 removes an item from the database, in a scenario where the application is deployed in only one region this would be as easy as remove one cache key, however this breaks down when deploying to multiple regions as user B in region eu-west-1 would still have the cache. B) Cache Indexes - In my projects I had a few instances where I would need to pre-calculate things and store them in the cache in order to be performant.

To summarise - while I see your point a DynamoDB global table would be really great, even for caching.

### Queues

As you said, a single SQS would be required, as all queue workers and functions in all regions would just read/write to it. As far as my knowable goes, SQS is a regional services anyway. This would probably not be an issues as most of the time, dispatched jobs in Laravel will wait for the response to be send before actually dispatching the job to the queue, so this will not cause a performance issue.

As far as the workers go, it would be great to either have workers in a single region or multiple. I think that multiple regions would be great for high availability, for example if us-east goes down we would still have the workers in eu-west-1.

TL:DR - SQS does not need to be deployed in multiple regions. It would be great to have workers in multiple regions.

### Scheduler

This is a bit tricky, we really only can deploy a single scheduler since it would execute multiple times when deployed in multiple regions. However this is also a single point of failure since when the region goes down in which the worker is deployed there is no backup.

Thanks for the response, I hope this helps.

**Reply by mnapoli** [reply]
Makes sense! The hard part might be the database of course. I would also strongly advise against deploying in multiple regions for redundancy: this adds a huge maintenance effort and complexity, for rare events (region being down) that may impact multiple regions anyway (e.g. us-east-1 can take down other regions too). It's always a tradeoff to consider (investment vs risk). But of course, it really depends on the project and its requirements. If you have a project right now and you want that we take a look at it together (I have some availability in the next weeks), feel free to reach out (matthieu@bref.sh) or book a bref enterprise intro call here: https://calendly.com/bref-enterprise/intro I'd be happy to learn more!
